framework/src/dirackernels/DiracKernel.C

```diff
@@ -193,139 +193,197 @@ DiracKernel::addPoint(const Elem * elem, Point p, unsigned /*id*/)
 const Elem *
 DiracKernel::addPoint(Point p, unsigned id)
 {
+  // Make sure that this method was called with the same id on all
+  // processors.  It's an extra communication, though, so let's only
+  // do it in DEBUG mode.
+  libmesh_assert(comm().verify(id));
+
   if (id != libMesh::invalid_uint)
+    return addPointWithValidId(p, id);
+
+  // If id == libMesh::invalid_uint (the default), the user is not
+  // enabling caching when they add Dirac points.  So all we can do is
+  // the PointLocator lookup, and call the other addPoint() method.
+  const Elem * elem = _dirac_kernel_info.findPoint(p, _mesh);
+  addPoint(elem, p, id);
+  return elem;
+}
+
+const Elem *
+DiracKernel::addPointWithValidId(Point p, unsigned id)
+{
+  // The Elem we'll eventually return.  We can't return early on some
+  // processors, because we may need to call parallel_only() functions in
+  // the remainder of this scope.
+  const Elem * return_elem = NULL;
+
+  // May be set if the Elem is found in our cache, otherwise stays as NULL.
+  const Elem * cached_elem = NULL;
+
+  // OK, the user gave us an ID, let's see if we already have it...
+  point_cache_t::iterator it = _point_cache.find(id);
+
+  // Was the point found in a _point_cache on at least one processor?
+  unsigned int i_found_it = static_cast<unsigned int>(it != _point_cache.end());
+  unsigned int we_found_it = i_found_it;
+  comm().max(we_found_it);
+
+  // If nobody found it in their local caches, it means we need to
+  // do the PointLocator look-up and update the caches.  This is
+  // safe, because all processors have the same value of we_found_it.
+  if (!we_found_it)
   {
-    // OK, the user gave us an ID, let's see if we already have it...
-    point_cache_t::iterator it = _point_cache.find(id);
-
-    // Was the point found in a _point_cache on at least one processor?
-    unsigned int i_found_it = static_cast<unsigned int>(it != _point_cache.end());
-    unsigned int we_found_it = i_found_it;
-    comm().max(we_found_it);
-
-    // If the point was found in a cache, but not my cache, I'm not responsible for it.
-    if (we_found_it && !i_found_it)
-      return NULL;
-
-    // Now that we only cache local data, some processors may enter
-    // this if statement and some may not.  Therefore we can't call
-    // any parallel_only() functions inside this if statement.
-    if (i_found_it)
+    const Elem * elem = _dirac_kernel_info.findPoint(p, _mesh);
+
+    // Only add the point to the cache on this processor if the Elem is local
+    if (elem && (elem->processor_id() == processor_id()))
```

Comment:
> a point literally falling through the cracks would surprise me.

I think it could maybe happen in mixed-dimension meshes, and in that case you the fix would probably be to specify a subdomain id to restrict the PointLocator to... but that interface is not currently available for the DiracKernel stuff.